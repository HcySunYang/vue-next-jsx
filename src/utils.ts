import * as bt from '@babel/types'

export function makeMap(
  str: string,
  expectsLowerCase?: boolean
): (key: string) => boolean {
  const map: Record<string, boolean> = Object.create(null)
  const list: Array<string> = str.split(',')
  for (let i = 0; i < list.length; i++) {
    map[list[i]] = true
  }
  return expectsLowerCase
    ? (val) => !!map[val.toLowerCase()]
    : (val) => !!map[val]
}

// https://developer.mozilla.org/en-US/docs/Web/HTML/Element
const HTML_TAGS =
  'html,body,base,head,link,meta,style,title,address,article,aside,footer,' +
  'header,h1,h2,h3,h4,h5,h6,hgroup,nav,section,div,dd,dl,dt,figcaption,' +
  'figure,picture,hr,img,li,main,ol,p,pre,ul,a,b,abbr,bdi,bdo,br,cite,code,' +
  'data,dfn,em,i,kbd,mark,q,rp,rt,rtc,ruby,s,samp,small,span,strong,sub,sup,' +
  'time,u,var,wbr,area,audio,map,track,video,embed,object,param,source,' +
  'canvas,script,noscript,del,ins,caption,col,colgroup,table,thead,tbody,td,' +
  'th,tr,button,datalist,fieldset,form,input,label,legend,meter,optgroup,' +
  'option,output,progress,select,textarea,details,dialog,menu,menuitem,' +
  'summary,content,element,shadow,template,blockquote,iframe,tfoot'

// https://developer.mozilla.org/en-US/docs/Web/SVG/Element
const SVG_TAGS =
  'svg,animate,animateMotion,animateTransform,circle,clipPath,color-profile,' +
  'defs,desc,discard,ellipse,feBlend,feColorMatrix,feComponentTransfer,' +
  'feComposite,feConvolveMatrix,feDiffuseLighting,feDisplacementMap,' +
  'feDistanceLight,feDropShadow,feFlood,feFuncA,feFuncB,feFuncG,feFuncR,' +
  'feGaussianBlur,feImage,feMerge,feMergeNode,feMorphology,feOffset,' +
  'fePointLight,feSpecularLighting,feSpotLight,feTile,feTurbulence,filter,' +
  'foreignObject,g,hatch,hatchpath,image,line,linearGradient,marker,mask,' +
  'mesh,meshgradient,meshpatch,meshrow,metadata,mpath,path,pattern,' +
  'polygon,polyline,radialGradient,rect,set,solidcolor,stop,switch,symbol,' +
  'text,textPath,title,tspan,unknown,use,view'

export const isHTMLTag = makeMap(HTML_TAGS)
export const isSVGTag = makeMap(SVG_TAGS)

// Transform `JSXMemberExpression` to `MemberExpression`
export function transformJSXMemberExpression(
  jsxMemberExpress: bt.JSXMemberExpression
): bt.MemberExpression {
  return bt.memberExpression(
    bt.isJSXIdentifier(jsxMemberExpress.object)
      ? bt.identifier(jsxMemberExpress.object.name)
      : transformJSXMemberExpression(jsxMemberExpress.object),
    bt.identifier(jsxMemberExpress.property.name)
  )
}

const cacheStringFunction = <T extends (str: string) => string>(fn: T): T => {
  const cache: Record<string, string> = Object.create(null)
  return ((str: string) => {
    const hit = cache[str]
    return hit || (cache[str] = fn(str))
  }) as any
}

const camelizeRE = /-(\w)/g
export const camelize = cacheStringFunction((str: string): string => {
  return str.replace(camelizeRE, (_, c) => (c ? c.toUpperCase() : ''))
})

export const capitalize = cacheStringFunction((str: string): string => {
  return str.charAt(0).toUpperCase() + str.slice(1)
})
