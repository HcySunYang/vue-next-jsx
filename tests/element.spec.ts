import { transformWithPlugin } from './utils'

describe('Element', () => {
  test('Normal element', () => {
    const code = `const el = <p></p>`
    transformWithPlugin(code)
  })

  test('Normal element with children', () => {
    const code = `
      const el = <p><i>hello</i> { ref.value }</p>
    `
    transformWithPlugin(code)
  })

  test('Normal element(self-closing)', () => {
    const code = `const el = <input id="foo"/>`
    transformWithPlugin(code)
  })

  test('Normal element with normal attribute', () => {
    const code = `const el = <p id="1" class='abc'></p>`
    transformWithPlugin(code)
  })

  test('Normal element with spread attribute', () => {
    const code = `const el = <div { ...attrs } id="foo" ></div>`
    transformWithPlugin(code)
  })

  test('Normal element with normal attribute', () => {
    const code = `const el = <p id="1" class='abc'></p>`
    transformWithPlugin(code)
  })

  test('Normal elements with children', () => {
    const code = `
    const el = (
      <div>
        <p>foo</p>
        TEXT
        { ...slots.bar() }
        { item }
      </div>
    )
    `
    transformWithPlugin(code)
  })
})
