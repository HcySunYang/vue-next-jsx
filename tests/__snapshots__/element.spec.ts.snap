// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element Normal element 1`] = `
"import { createVNode } from \\"vue\\";
const el = createVNode(\\"p\\");"
`;

exports[`Element Normal element with children 1`] = `
"import { createTextVNode, createVNode } from \\"vue\\";
const el = createVNode(\\"p\\", null, [createVNode(\\"i\\", null, [createTextVNode(\\"hello\\")]), createTextVNode(\\" \\"), ref.value]);"
`;

exports[`Element Normal element with normal attribute 1`] = `
"import { createVNode } from \\"vue\\";
const el = createVNode(\\"p\\", {
  \\"id\\": \\"1\\",
  \\"class\\": 'abc'
});"
`;

exports[`Element Normal element with normal attribute 2`] = `
"import { createVNode } from \\"vue\\";
const el = createVNode(\\"p\\", {
  \\"id\\": \\"1\\",
  \\"class\\": 'abc'
});"
`;

exports[`Element Normal element with spread attribute 1`] = `
"import { mergeProps, createVNode } from \\"vue\\";
const el = createVNode(\\"div\\", mergeProps({ ...attrs
}, {
  \\"id\\": \\"foo\\"
}));"
`;

exports[`Element Normal element(self-closing) 1`] = `
"import { createVNode } from \\"vue\\";
const el = createVNode(\\"input\\", {
  \\"id\\": \\"foo\\"
});"
`;

exports[`Element Normal elements with children 1`] = `
"import { createTextVNode, createVNode } from \\"vue\\";
const el = createVNode(\\"div\\", null, [createVNode(\\"p\\", null, [createTextVNode(\\"foo\\")]), createTextVNode(\\"TEXT\\"), ...slots.bar(), item]);"
`;
